### PR TITLE
Add test for sycl::has_kernel_bundle from multiple kernels test case

### PR DESCRIPTION
**Pull request by vasilytric**
Provide a set of tests that verify logic of multiple kernel ids usage in comparison to single kernel id usage.

Specifically check that calling `has_kernel_bundle` overload for a multiple kernels gives the same result as:
* call has_kernel_bundle overload for each of kernels and
* apply either `AND` or `OR`  operation over the results of the separate calls
  * for `has_kernel_bundle(ctx)`: the `OR` operation
  * for `has_kernel_bundle(ctx, devices)`: the `OR` operation
  * for `has_kernel_bundle(ctx, kernel_ids)`: the `AND` operation
  * for `has_kernel_bundle(ctx, devices, kernel_ids)`: the `AND` operation

___________________________________________
This PR depend on these PRs:
- https://github.com/KhronosGroup/SYCL-CTS/pull/218
- https://github.com/KhronosGroup/SYCL-CTS/pull/220

### CODE REVIEW

**Review by ProGTX — APPROVED** [review]

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev_and_k_id.cpp [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+//  SYCL 2020 Conformance Test Suite
+//
+//  This test calls sycl::has_kernel_bundle<all bundle_states>(ctx, devices,
+//  kernelId) for kernels with this six kernels (2x kernel_cpu, 2x kernel_gpu,
+//  2x kernel_accelerator).
+//    1) Call sycl::has_kernel_bundle only with one kernel: with kernel_cpu,
+//  kernel_gpu and kernel_accelerator.
+//    2) Unite call results for two kernels, for example call result for
+//  kernel_cpu and kernel_gpu (union using && operator). Save union result to
+//  varaible with name "expected"
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev_and_k_id.cpp]: I have a grammar spell-checker, but in some cases were missed by myself
This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-4abdb25438ae5e8dc1668e5995298ef5a9776db541db27c6fd9ab5f126c27e0bR11)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev_and_k_id.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  This test calls sycl::has_kernel_bundle<all bundle_states>(ctx, devices,
+//  kernelId) for kernels with this six kernels (2x kernel_cpu, 2x kernel_gpu,
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev_and_k_id.cpp]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-4abdb25438ae5e8dc1668e5995298ef5a9776db541db27c6fd9ab5f126c27e0bR6)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev.cpp [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+//  unite each call result to varaible with name "expected" (unition using "|="
+//  operator), then calls sycl::has_kernel_bundle<sycl::bndle_state>(context,
+//  devices).
+//
+//  The test verifies that variable value "expected" is equal to
+//  sycl::has_kernel_bundle<sycl::bundle_state>(context, devices) calling result
+//
+//  IMPORTANT: This file should not be merged with another *.cpp file because we
+//  we must have a strictly defined kernels in the application because tested
+//  overload uses all available kernels and we can't chose specific kernel.
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev.cpp:22]: Thanks :)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev.cpp [hunk 4]
```diff
@@ -0,0 +1,56 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  The test calls sycl::has_kernel_bundle<sycl::bundle_state>(context, device,
+//  kernelId) with each one of all available kernels in the application and
+//  unite each call result to varaible with name "expected" (unition using "|="
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels_dev.cpp]: Yes, comments in `has_kernel_bundle_multiple_kernels_dev` and `has_kernel_bundle_multiple_kernels_ctx_only` tests were reworked.
But this comment ("unition using "|=" ...") is not actual...
This comment was reworked [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-a9f43a476db014320dc33441820351300bf3a62d04263b92b334d2619585a56bR5-R15)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels_ctx_only.cpp [hunk 5]
```diff
@@ -0,0 +1,55 @@
+/*******************************************************************************
+//
+//  SYCL 2020 Conformance Test Suite
+//
+//  The test calls sycl::has_kernel_bundle<sycl::bundle_state>(context, device,
+//  kernelId) with each one of all available kernels in the application and
+//  unite each call result to varaible with name "expected" (unition using "|="
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels_ctx_only.cpp]: This comment was reworked [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-c5b8cc34f62ee7efc904627b7b236d54f9d655cc7fd337f2cc4423c750cd3fa4R5-R15)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h [hunk 6]
```diff
@@ -0,0 +1,306 @@
...
+
+/** @brief Specialization for case of has_kernel_bundle(context) overload
+ */
+template <sycl::bundle_state BundleState>
+struct multiple_kernels<BundleState, overload::id::ctx_only> {
+  static void run(sycl_cts::util::logger &log, const sycl::device &device) {
+    const std::vector<sycl::device> devices{device};
+    sycl::context ctx(devices);
+
+    auto call = [&]() { return sycl::has_kernel_bundle<BundleState>(ctx); };
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-46ece85fa8dc115567742a60a4e04a764d58ba6c59d7c40b3bcfc0cba7166d9cL298-L299)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h [hunk 7]
```diff
@@ -0,0 +1,306 @@
...
+/** @brief Specialization for case of has_kernel_bundle(context, devices)
+ *         overload
+ */
+template <sycl::bundle_state BundleState>
+struct multiple_kernels<BundleState, overload::id::ctx_dev> {
+  static void run(sycl_cts::util::logger &log, const sycl::device &device) {
+    const std::vector<sycl::device> devices{device};
+    sycl::context ctx(devices);
+
+    auto call = [&]() {
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-46ece85fa8dc115567742a60a4e04a764d58ba6c59d7c40b3bcfc0cba7166d9cL282-L283)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h [hunk 8]
```diff
@@ -0,0 +1,306 @@
...
+    const auto ctx_devices = ctx.get_devices();
+    if (ctx_devices != devices) {
+      FAIL(log, "Pre-condition failed, test issue");
+    }
+  }
+
+  // Retrieve results of a single calls
+  bool expected = false;
+  for (const auto &id : ids) {
+    expected |= sycl::has_kernel_bundle<BundleState>(ctx, devices, {id});
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-46ece85fa8dc115567742a60a4e04a764d58ba6c59d7c40b3bcfc0cba7166d9cR217-R219)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h [hunk 9]
```diff
@@ -0,0 +1,306 @@
...
+
+    if (result != expected) {
+      on_failure(result, expected, "two kernels with CPU aspects");
+    }
+  }
+}
+
+/** @brief Check that calling has_kernel_bundle overload for a two kernels
+ *         is same as:
+ *           - call has_kernel_bundle overload for each of kernels and
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-46ece85fa8dc115567742a60a4e04a764d58ba6c59d7c40b3bcfc0cba7166d9cR158-R160)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h [hunk 10]
```diff
@@ -0,0 +1,306 @@
...
+ *           - apply 'AND' operation over the results of the separate calls
+ *  @details According to the SYCL-2020 rev.3 #4.11.8 "Querying if a kernel
+ *           bundle exists" we need to ensure that
+ *           - if we call specific sycl::has_kernel_bundle overload for multiple
+ *             kernel ids
+ *           - the result "true" is expected only if every kernel that used
+ *             is compatible with device
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-46ece85fa8dc115567742a60a4e04a764d58ba6c59d7c40b3bcfc0cba7166d9cR30-R31)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h [hunk 11]
```diff
@@ -0,0 +1,306 @@
...
+#include "has_kernel_bundle.h"
+#include "kernels.h"
+#include <string>
+#include <vector>
+
+namespace sycl_cts::testshas_kernel_bundle::check {
+namespace details {
+/** @brief Check that calling has_kernel_bundle overload for a two kernels
+ *         is same as:
+ *           - call has_kernel_bundle overload for each of kernels and
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-46ece85fa8dc115567742a60a4e04a764d58ba6c59d7c40b3bcfc0cba7166d9cR24)

File: tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h [hunk 12]
```diff
@@ -0,0 +1,306 @@
...
+#include "../common/common.h"
+#include "has_kernel_bundle.h"
+#include "kernels.h"
+#include <string>
+#include <vector>
+
+namespace sycl_cts::testshas_kernel_bundle::check {
+namespace details {
+/** @brief Check that calling has_kernel_bundle overload for a two kernels
+ *         is same as:
```
[the remark this answers is not in the document]
Comment (vasilytric) [reply tests/kernel_bundle/has_kernel_bundle_multiple_kernels.h]: This comment was applied [here](https://github.com/KhronosGroup/SYCL-CTS/pull/222/commits/6e271ac73e2351992d0e68a8ce859568f7057671#diff-46ece85fa8dc115567742a60a4e04a764d58ba6c59d7c40b3bcfc0cba7166d9cR23)

**Pull request merged**